lib/puppet/ssl/host.rb

```diff
@@ -371,27 +382,67 @@ def state
 
   private
 
+  # Ensures that CRL is either on disk, or that CRL checking has been disabled
+  def ensure_crl_if_needed
+    if use_crl? && !Puppet::FileSystem.exist?(crl_path)
+      # The CertificateRevocationList indirector will attempt to download the
+      # CRL from the CA if it does not exist on disk. It will ask host for
+      # its ssl_store again, but expect crl checking to be disabled for that
+      # store.
+      old_crl_usage = @crl_usage
+      @crl_usage = false
+      Puppet.debug _("Disabling certificate revocation checking when fetching the CRL and no CRL is present")
+      if !CertificateRevocationList.indirection.find(CA_NAME)
+        raise Puppet::Error,
+              _("Certificate revocation checking is enabled but a CRL cannot be found; CRL checking will not be performed.")
+      end
+      @crl_usage = old_crl_usage
+    end
+  end
+
+  # @param path [String] Path to CRL Chain
+  # @return [Array<OpenSSL::X509::CRL>] CRLs from chain
+  # @raise [Errno::ENOENT] if file does not exist
+  # @raise [Puppet::Error<OpenSSL::X509::CRLError>] if the CRL chain is malformed
+  def load_crls(path)
+    ending = "-----END X509 CRL-----\n"
+    crls_pems = Puppet::FileSystem.read(path, encoding: Encoding::ASCII)
```

Comment:
Might be worth looking to see if this should be UTF-8; other parts of the [code base](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/ssl/configuration.rb#L53) suggest that comments are allowed per the [RFC](https://www.ietf.org/rfc/rfc2459.txt).

Reply:
I think I made a comment at some point that I was writing files as ASCII in my PR, but I take it back, I'm writing as UTF-8.

Reply:
Looking into the RFCs, it looks like UTF8 is valid in a number of cert fields (names and uris) that could also be put into a CRL (I didn't find a mention of UTF8 comments in that RFC though). However, we have the requirement that these are concatenated PEM encoded objects, which have ASCII headers and footers with base64 bodies. However, the twist gets twistier, the PEM standard allows for "[Explanatory Text](https://tools.ietf.org/html/rfc7468#section-5.2)" and their example includes fields from a cert that could include UTF8.

So, I think that UTF-8 encoding should be the way to go since the PEM encoding will be a UTF8 compatible subset of ASCII and there could be explanatory text with UTF8.

Also, since there could be explanatory text, I think the String#scan + regex matching that the code Tony linked to will get us just the PEM encoding without any potential explanatory text.